File: tasks/implementations/php/tasks/web_scraper.php
<?php
function run_web_scraper(string $size, string $fixturesRoot): void {
  $base = $fixturesRoot . "/mock_site/{$size}";
  $index = file_get_contents($base . '/index.html');
  preg_match_all('/href="([^"]+)"/', $index, $matches);
  $total = 0;
  foreach ($matches[1] as $idx => $link) {
    $page = file_get_contents($base . '/' . $link);
    preg_match('/<h2>(.*?)<\/h2>/', $page, $title);
    preg_match('/<p>(.*?)<\/p>/', $page, $body);
    $total += ($idx + 1) * (strlen($title[1]) + strlen($body[1]));
  }
  echo $total . PHP_EOL;
}

File: tasks/implementations/rust/tasks/web_scraper.rs
use std::fs;

pub fn run(size: &str, fixtures_root: &str) {
    let base = format!("{}/mock_site/{}", fixtures_root, size);
    let index = fs::read_to_string(format!("{}/index.html", base)).unwrap();
    let mut total = 0i64;
    let mut idx = 0i64;
    for link in index.match_indices("href=\"") {
        let start = link.0 + 6;
        let rest = &index[start..];
        let end = rest.find('"').unwrap();
        let path = &rest[..end];
        let page = fs::read_to_string(format!("{}/{}", base, path)).unwrap();
        let title = between(&page, "<h2>", "</h2>");
        let body = between(&page, "<p>", "</p>");
        idx += 1;
        total += idx * (title.len() as i64 + body.len() as i64);
    }
    println!("{}", total);
}

fn between<'a>(text: &'a str, left: &str, right: &str) -> &'a str {
    let start = text.find(left).unwrap() + left.len();
    let rest = &text[start..];
    let end = rest.find(right).unwrap();
    &rest[..end]
}
